### PR TITLE
[WIP] Replace watcher with Broccoli watcher

### PR DESCRIPTION
**Pull request by oligriffiths**
This PR is an attempt to replace the Ember CLI file watcher with the Broccoli one.
At this point I'm only replacing the watcher, not the server etc to keep the changes contained, but in the future I see these also moving to Broccoli.

There are quite a few differences in the API and functionality of the Ember-CLI watcher and the Broccoli one.
The Ember version https://github.com/ember-cli/broccoli-sane-watcher/blob/ember-cli/index.js#L116 sets up a bunch of handlers on the build that do the following:

```js
  return this.builder
    .build(addWatchDir, annotation) // 1
    .then(saveNode) // 2
    .then(totalTime) // 3
    .then(appendFilePath) // 4
    .then(triggerChange, triggerError) // 5
    .then(verboseOutput.bind(this)) // 6
    .finally(cleanup); // 2
```

1. Notify the watcher of watched directories - this isn't necessary as Broccoli watcher does it a different way https://github.com/broccolijs/broccoli/blob/master/lib/watcher.js#L39
2. Some heimdall cleanup
3. Total time calculations (weirdly this total time is different from the output nodes totalTime stat...)
4. `appendFilePath` which appears to do nothing as `filePath` is null https://github.com/ember-cli/broccoli-sane-watcher/blob/ember-cli/index.js#L58
5. Triggers a change in the Ember watcher model
6. Print slow nodes

All of the above (except 4) have been ported into the Ember watcher model.
These changes are needed in broccoli https://github.com/broccolijs/broccoli/pull/398 so will land that first.

### CODE REVIEW

File: lib/models/watcher.js [hunk 1]
```diff
@@ -4,44 +4,138 @@ const chalk = require('chalk');
...
+        break;
+      case 'delete':
+        action = 'deleted';
+        break;
+      default:
+        action = event;
+        break;
+    }
+
+    console.log(`file ${action} ${filePath}`);
```
Comment (stefanpenner) [inline lib/models/watcher.js]: can we use the `ui` object to write to stdout?

File: lib/models/watcher.js [hunk 2]
```diff
@@ -4,44 +4,138 @@ const chalk = require('chalk');
...
     this.analytics.trackError({
       description: error && error.name,
     });
   }
 
-  then() {
-    return this.watcher.then.apply(this.watcher, arguments);
+  _didChange(event, filePath) {
+    let action;
+    switch (event) {
```
Comment (stefanpenner) [inline lib/models/watcher.js]: nit: Maybe a lookup table would be nicer?

**Comment by stefanpenner** [comment]
@oligriffiths can we target this to the beta branch?

**Comment by stefanpenner** [comment]
@oligriffiths nice thanks!

File: lib/models/watcher.js [hunk 3]
```diff
@@ -16,15 +18,88 @@ class Watcher extends CoreObject {
...
+    if (isExperimentEnabled('BROCCOLI_WATCHER')) {
+      this.watcher = this.watcher || this._constructBroccoliWatcher(options);
+      this.watcher.on('change', this.didChange.bind(this));
+      this.watcher.on('buildSuccess', this.didBuild.bind(this));
+      this.watcher.start();
+      this._setupBuild(this.watcher.currentBuild);
+    } else {
+      this.watcher = this.watcher || this._constructWatcher(options);
+      this.watcher.on('change', this.didBuild.bind(this));
+      this.build = this.watche;
```
Comment (stefanpenner) [inline lib/models/watcher.js]: This looks incomplete, also not setting `this.build` symmetrically makes this slightly harder to read.

File: lib/models/watcher.js [hunk 4]
```diff
@@ -16,15 +18,88 @@ class Watcher extends CoreObject {
...
     this.watcher.on('error', this.didError.bind(this));
-    this.watcher.on('change', this.didChange.bind(this));
     this.serveURL = serveURL;
   }
 
-  constructWatcher(options) {
-    return new (require('ember-cli-broccoli-sane-watcher'))(this.builder, options);
+  _constructWatcher(options) {
+    const watcher = new (require('ember-cli-broccoli-sane-watcher'))(this.builder, options);
+    this.build = watcher;
```
Comment (stefanpenner) [inline lib/models/watcher.js]: `_constructWatcher` having side-affects, by placing `watcher` on `this.build`, is surprising. It should likely be symmetrical with where else `this.build` is set.

**Comment by stefanpenner** [comment]
rebased + tests fixed -> https://github.com/ember-cli/ember-cli/pull/8666#pullrequestreview-244071079

**Comment by lifeart** [comment]
on windows:

```
ERROR Summary:

  - broccoliBuilderErrorStack: [undefined]
  - code: [undefined]
  - codeFrame: [undefined]
  - errorMessage: this.watcher.currentBuild.then(...).finally is not a function
  - errorType: [undefined]
  - location:
    - column: [undefined]
    - file: [undefined]
    - line: [undefined]
  - message: this.watcher.currentBuild.then(...).finally is not a function
  - name: TypeError
  - nodeAnnotation: [undefined]
  - nodeName: [undefined]
  - originalErrorMessage: [undefined]
  - stack: TypeError: this.watcher.currentBuild.then(...).finally is not a function
    at Watcher._setupBroccoliWatcherBuild (C:\Users\Aleksandr_Kanunnikov\Documents\repos\ember-broccoli-memo-debug\node_modules\ember-cli\lib\models\watcher.js:59:15)
    at emitNone (events.js:106:13)
    at Watcher.emit (events.js:208:7)
    at Watcher._build (C:\Users\Aleksandr_Kanunnikov\Documents\repos\ember-broccoli-memo-debug\node_modules\ember-cli\node_modules\broccoli\lib\watcher.js:96:10)
    at currentBuild.Promise.resolve.then.then (C:\Users\Aleksandr_Kanunnikov\Documents\repos\ember-broccoli-memo-debug\node_modules\ember-cli\node_modules\broccoli\lib\watcher.js:50:34)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)

=================================================================================

```

**Comment by lifeart** [comment]
fix: `watcher.js`
const promiseFinally = require('promise.prototype.finally');


```js
  _setupBroccoliWatcherBuild() {
    let heimdallNode;

    promiseFinally(this.watcher.currentBuild
      .then(hash => {
        heimdallNode = hash.graph.__heimdall__;
        return hash;
      }),() => {
        // guard against `build` rejecting
        if (heimdallNode) {
          // remove the heimdall subtree for this build so we don't leak.  If
          // BROCCOLI_VIZ=1 then we have already output the json in `verboseOutput`.
          heimdallNode.remove();
        }
      });
  }
```

**Comment by rwjblue** [comment]
@lifeart - Mind filing an issue for that? Its really easy to miss comments on closed PR's / issues like this.

**Comment by stefanpenner** [comment]
I'll fix.

**Comment by stefanpenner** [comment]
https://github.com/ember-cli/ember-cli/pull/8674

**Pull request closed without merging**